### PR TITLE
feat(sock): Add a new cancellation callback interface.

### PR DESCRIPTION
**Pull request by royjacobson**
There's a minimal Dragonfly PR ready as well.

### CODE REVIEW

File: util/fibers/epoll_socket.cc [hunk 1]
```diff
@@ -45,7 +45,7 @@ nonstd::unexpected<error_code> MakeUnexpected(std::errc code) {
 }
 
 #ifdef __linux__
-constexpr int kEventMask = EPOLLIN | EPOLLOUT | EPOLLET;
+constexpr int kEventMask = EPOLLIN | EPOLLOUT | EPOLLET | EPOLLHUP | EPOLLERR | EPOLLRDHUP;
```
Comment (romange) [inline util/fibers/epoll_socket.cc:48]: see https://man7.org/linux/man-pages/man2/epoll_ctl.2.html

HUP/ERR are always reported

File: util/fibers/uring_socket.h [hunk 2]
```diff
@@ -60,7 +63,9 @@ class UringSocket : public LinuxSocketBase {
   uint8_t register_flag() const {
     return fd_ & REGISTER_FD ? IOSQE_FIXED_FILE : 0;
   }
+
+  uint32_t error_cb_event_ = UINT32_MAX;
```
Comment (romange) [inline util/fibers/uring_socket.h:67]: maybe call it `error_cb_id_` ?  event can be confused with the eventmask.

File: util/fibers/epoll_socket.cc [hunk 3]
```diff
@@ -426,7 +436,7 @@ bool EpollSocket::SuspendMyself(detail::FiberInterface* cntx, std::error_code* e
 void EpollSocket::Wakey(uint32_t ev_mask, int error, EpollProactor* cntr) {
   DVSOCK(2) << "Wakey " << ev_mask;
 #ifdef __linux__
-  constexpr uint32_t kErrMask = EPOLLERR | EPOLLHUP;
+  constexpr uint32_t kErrMask = EPOLLERR | EPOLLHUP | EPOLLRDHUP;
```
Comment (romange) [inline util/fibers/epoll_socket.cc:439]: was there a reason for EPOLLRDHUP ? I think we got the event with `EPOLLHUP` as well.

Comment (royjacobson) [reply util/fibers/epoll_socket.cc:439]: I get a `0x2005` aka `EPOLLIN | EPOLLOUT | EPOLLRDHUP` mask when the socket closes.

**Review by romange — APPROVED** [review]

**Pull request merged**